### PR TITLE
ENH: make is_list_like handle non iterable numpy-like arrays correctly

### LINKED ISSUE #35131: ENH: Identify numpy-like zero-dimensional arrays as non-iterable
Note: I wasn't sure whether to label this as an enhancement or bug. I've put enhancement as I don't think the original implementation was actually incorrect, rather just limited in scope. I'm happy to change to bug if that's more appropriate.

#### Is your feature request related to a problem?

I'm one of the developers of [Pint-Pandas](https://github.com/hgrecco/pint-pandas), which provides an extension type for storing numeric data with units (and, once mature, may solve #10349). The problem we're having is that we can't pass all the extension type tests. In particular, we fail many tests due to Panda's current implementation of `is_list_like`. 

As @hgrecco shows [here](https://github.com/hgrecco/pint-pandas/issues/33#issuecomment-647225205), pandas determines whether something `is_list_like` [here](https://github.com/pandas-dev/pandas/blob/4743cc91719146167eefb43865413fc0b0aef1d4/pandas/_libs/lib.pyx#L992) (code copied below).

```python
cdef inline bint c_is_list_like(object obj, bint allow_sets) except -1:
    return (
        isinstance(obj, abc.Iterable)
        # we do not count strings/unicode/bytes as list-like
        and not isinstance(obj, (str, bytes))
        # exclude zero-dimensional numpy arrays, effectively scalars
        and not (util.is_array(obj) and obj.ndim == 0)
        # exclude sets if allow_sets is False
        and not (allow_sets is False and isinstance(obj, abc.Set))
    )
```

At the moment, all Pint `Quantity` instances have an `__iter__` method. It is designed to raise a `TypeError` if the quantity isn't actually iterable (code [here](https://github.com/hgrecco/pint/blob/4ab68ba0a00f96ed68fdf084233b61d68127c461/pint/quantity.py#L242)). 

The problem is that the `__iter__` method is defined, regardless of whether the `Quantity` instance is actually iterable or not. This is a problem because it means that, in all cases, `isinstance(quantity_instance, abc.Iterable)` returns `True` (even if `quantity_instance` isn't actually iterable and would return a `TypeError` as soon as one tried to iterate with it). 

A numpy array suffers from the same issue which is where the `not (util.is_array(obj) and obj.ndim == 0)` in Panda's `is_list_like` is important, because it excludes scalars. The problem is that `util.is_array(obj)` returns `False` if `obj` is anything other than `np.ndarray`. As a result, we can't just add an `ndim` attribute to `Quantity` because the `obj.ndim == 0` check will never be evaluated anyway. 

#### Describe the solution you'd like

Update `is_list_like` so that it correctly identifies implementations of scalars in numpy-like arrays (I think these are called 'duck numpy arrays'?). I've put a draft in #35127 and would be grateful for feedback on the proposed solution there. If there's a better solution, I'm also happy to implement that instead. 

#### API breaking implications

`is_list_like` would now return `False` for scalars of numpy-like array implementations. If we used the implementation in #35127 then any object that has an `ndim` attribute and has `obj.ndim == 0` would no longer be identified as list-like. 

#### Describe alternatives you've considered

We've discussed how we could get the behaviour we want by changing Pint, rather than changing Pandas, but it seems to be impossible if we want to maintain the numpy-like implementation without separate scalar and array classes (see discussion [here](https://github.com/hgrecco/pint/issues/1128)).

#### Additional context

```python
>>> import numpy as np
>>> import pandas as pd
>>> import pint
>>> from collections import abc

>>> numpy_scalar = np.array(2)
>>> numpy_scalar
array(2)
# numpy 0d-array (i.e. scalar) quantities appear to be iterable even though they're not
>>> isinstance(numpy_scalar, abc.Iterable)
True
>>> numpy_scalar[0]
...
IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed

# however, numpy 0d-array quantities are correctly handled by Pandas
>>> pd.core.dtypes.inference.is_list_like(numpy_scalar)
False

# compare this with Pint's numpy-like implementation
>>> ureg = pint.UnitRegistry()
>>> quantity_scalar = 3 * ureg.meter
>>> quantity_scalar
<Quantity(3, 'meter')>

# similar to a numpy array, scalar quantities appear to be iterable even though they're not
>>> isinstance(quantity_scalar, abc.Iterable)
True
>>> quantity_scalar[0]  
...
TypeError: Neither Quantity object nor its magnitude (3)supports indexing

# however, Pandas identifies Pint's implementation as being iterable which is what is causing
# our problem
>>> pd.core.dtypes.inference.is_list_like(quantity_scalar)
True  # we would like this to be False
```

### PR DESCRIPTION
**Pull request by znicholls**
- [x] closes #35131
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -994,8 +994,8 @@ cdef inline bint c_is_list_like(object obj, bint allow_sets) except -1:
         isinstance(obj, abc.Iterable)
         # we do not count strings/unicode/bytes as list-like
         and not isinstance(obj, (str, bytes))
-        # exclude zero-dimensional numpy arrays, effectively scalars
-        and not (util.is_array(obj) and obj.ndim == 0)
+        # exclude zero-dimensional duck arrays, effectively scalars
+        and not (hasattr(obj, "ndim") and obj.ndim == 0)
```
Comment (keewis) [inline pandas/_libs/lib.pyx:998]: you might be able to simplify this using `getattr`:
```suggestion
        and getattr(obj, "ndim", 0) != 0
```

Comment (znicholls) [reply pandas/_libs/lib.pyx]: Awesome thanks!

**Comment by jbrockmendel** [comment]
This is likely to have a non-trivial effect on performance.  It might be time to implement a separate is_list_like (and possibly is_scalar) with less-strict/EA-customizable behavior (cc @jorisvandenbossche I know this has come up w/r/t geopandas)

Can you identify a subset of is_list_like usages where we need the less-strict behavior?

**Comment by znicholls** [comment]
> Can you identify a subset of is_list_like usages where we need the less-strict behavior?

I can check with the other pint-pandas developers. 

Can I also ask which [benchmarks](https://pandas.pydata.org/docs/dev/development/contributing.html#running-the-performance-test-suite) I should look at to examine whether performance has taken a hit?

**Comment by keewis** [comment]
> I'm not completely sure why, but reverting here for simplicity

yeah, sorry, `getattr` only checks for `AttributeError` but `pint` raises `TypeError`

**Comment by znicholls** [comment]
> yeah, sorry, `getattr` only checks for `AttributeError` but `pint` raises `TypeError`

I think it would have to be `getattr(obj, "ndim", 1) != 0` i.e. you assume everything is iterable unless you have evidence otherwise.

**Comment by znicholls** [comment]
I think the test failures are unrelated to our change...

```
...
WARNING: autodoc: failed to import method 'sparse.from_spmatrix' from module 'DataFrame'; the following exception was raised:
No module named 'DataFrame'
WARNING: autodoc: failed to import method 'sparse.to_coo' from module 'DataFrame'; the following exception was raised:
No module named 'DataFrame'
WARNING: autodoc: failed to import method 'sparse.to_dense' from module 'DataFrame'; the following exception was raised:
No module named 'DataFrame'
WARNING: autodoc: failed to import method 'sparse.from_coo' from module 'Series'; the following exception was raised:
No module named 'Series'
WARNING: autodoc: failed to import method 'sparse.to_coo' from module 'Series'; the following exception was raised:
No module named 'Series'
...
```

**Comment by jorisvandenbossche** [comment]
> This is likely to have a non-trivial effect on performance. 

To have an idea about this, I timed the following cases:

```
a1 = [1, 2, 3]
a2 = np.array([1, 2, 3])
a3 = np.array([1, 2, 3])[0] 
```

```
In [2]: %timeit pd.api.types.is_list_like(a1)      
390 ns ± 1.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)   # master  
553 ns ± 1.77 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)   # PR

In [3]: %timeit pd.api.types.is_list_like(a2) 
405 ns ± 0.972 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)   # master
408 ns ± 0.745 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)   # PR

In [4]: %timeit pd.api.types.is_list_like(a3)  
395 ns ± 1.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)   # master
411 ns ± 1.55 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)    PR
```

So for numpy arrays the slowdown is only very limited (I assume because in this case we needed to look up `obj.ndim` anyway, so then then `getattr(obj, "ndim")` seems to be similar). For actual lists, the slowdown is more significant, though (but I would think the array cases are most relevant for performance).

**Comment by jorisvandenbossche** [comment]
> It might be time to implement a separate is_list_like (and possibly is_scalar) with less-strict/EA-customizable behavior

It's not only EAs, though, but eg also for object dtype (see https://github.com/pandas-dev/pandas/issues/26333, https://github.com/pandas-dev/pandas/issues/27911)

But indeed, a less strict `is_list_like` check depending on the dtype of the involved data might be worth exploring, although not that easy I think. The first thing would be to identify the cases where this less strict case is actually important. I suppose the problems in pint-pandas mainly occur in indexing + arithmetic/comparison operations ?

**Comment by andrewgsavage** [comment]
These were the tests which started passing when deleting the `__iter__` method from the scalar class to make is_list_like return False:

<details>
TestGroupby::test_groupby_apply_identity
TestMethods::test_fillna_copy_frame
TestMethods::test_fillna_copy_series
TestMissing::test_fillna_scalar
TestMissing::test_fillna_series
TestMissing::test_fillna_frame
TestReshaping::test_unstack[series-index1]
TestReshaping::test_unstack[series-index3]
TestReshaping::test_unstack[frame-index1]
TestReshaping::test_unstack[frame-index3]
TestSetitem::test_setitem_mask_broadcast[loc]
TestSetitem::test_setitem_sequence_broadcasts[True]
TestSetitem::test_setitem_integer_array[True-list]
TestSetitem::test_setitem_integer_array[True-integer-array]
TestSetitem::test_setitem_integer_array[True-numpy-array]
TestSetitem::test_setitem_slice[True]
TestSetitem::test_setitem_loc_iloc_slice
</details>

**Comment by znicholls** [comment]
It looks like using numpy iterable speeds up the list and array cases but not the numpy scalar one (which I think makes sense as you have to do some more thinking on that one). (This comment is pending tests passing...)

```python
import numpy as np
import pandas as pd

a1 = [1, 2, 3]
a2 = np.array([1, 2, 3])
a3 = np.array([1, 2, 3])[0]


%timeit pd.api.types.is_list_like(a1)
289 ns ± 16.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # PR
447 ns ± 7.82 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master

%timeit pd.api.types.is_list_like(a2)
302 ns ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # PR
466 ns ± 4.96 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master

%timeit pd.api.types.is_list_like(a3)
596 ns ± 2.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # PR
440 ns ± 1.68 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master
```

**Comment by znicholls** [comment]
> Can you identify a subset of is_list_like usages where we need the less-strict behavior?

Having dug a bit deeper, we're going to need to back to the drawing board on this. My impression is that in order to be handled by `is_list_like` and `is_scalar` correctly, the class has to basically be a numpy array (or sub-class thereof). I'll go back to the pint-pandas devs and we can discuss how we go forward there, including identifying the cases where a less strict `is_list_like` and `is_scalar` would solve our problems. 

I'm happy to close this in the meantime. However it would be great if we could reopen #22582 to at least add pint-pandas to the docs (buggy as it is).

**Comment by hgrecco** [comment]
I would be nice to talk to geopandas devs to see if we can join forces. We might be facing the same problem.

**Comment by jorisvandenbossche** [comment]
@hgrecco you are alread talking to them ;) (well, at least, I am one of the geopandas devs) 
See also my comment at https://github.com/pandas-dev/pandas/issues/35131#issuecomment-654053344: although very much related, I think it's not exactly the same. As for GeoPandas, the scalar elements we put in a Series are actually iterable (in contrast to the (duck-typed) numpy scalars, which are iterable in theory (have the dunder method), but raise once you try to iterate).

**Comment by jorisvandenbossche** [comment]
> It looks like using numpy.iterable speeds up the list and array cases

It seems that `np.iterable` is actually just a try/except of `iter(value)`:

```
    try:
        iter(y)
    except TypeError:
        return False
    return True
```

(if we decide to use this, I would prefer doing the code above explicitly instead of using `np.iterable`, as I think most people are not familiar with that numpy method, and the code is simple).

But so basically it is a choosing between `isinstance(value, Iterable)` + special case check to exclude `ndim==0` vs try/except of `iter(value)` (not fully sure if there are other consequences of the choice between both)

**Comment by hgrecco** [comment]
@jorisvandenbossche thanks for the clarification. 

@znicholls Have you tried reinstating `getattr(obj, "ndim", 1) != 0` and moving it first? Python's and/or are shortcircuit operators so we might want to put first what we think it might fail first.

**Comment by znicholls** [comment]
Reverting the changes as suggested by @hgrecco leads to a slowdown for lists, no change for numpy arrays and a speedup for numpy scalars (essentially the reverse of previously)

```python
import numpy as np
import pandas as pd

a1 = [1, 2, 3]
a2 = np.array([1, 2, 3])
a3 = np.array([1, 2, 3])[0]


%timeit pd.api.types.is_list_like(a1)
640 ns ± 6.82 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # PR
447 ns ± 7.82 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master

%timeit pd.api.types.is_list_like(a2)
414 ns ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # PR
466 ns ± 4.96 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master

%timeit pd.api.types.is_list_like(a3)
385 ns ± 10.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) #PR
440 ns ± 1.68 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each) # master
```

**Comment by znicholls** [comment]
We still have the `is_scalar` and `item_from_zerodim` issues of course...

**Comment by znicholls** [comment]
@jbrockmendel I went through and identified where different `is_list_like` and `is_scalar` would be needed, see https://github.com/pandas-dev/pandas/compare/master...znicholls:handle-extension-arrays. Is that helpful?

**Comment by znicholls** [comment]
@jorisvandenbossche and @jbrockmendel do you have a sense of how best to proceed here? My impression is that this will require two, if not three, steps:

1. decide where we would like to use less strict implementations of `is_list_like` and `is_scalar` (see https://github.com/pandas-dev/pandas/compare/master...znicholls:handle-extension-arrays for where changes are needed for our Pint extension array implementation to work)
1. implement less strict `is_list_like` and `is_scalar` in the internals where there is no performance issue
1. (with luck, not a necessary step) come up with solutions for places where we'd like to have less strict implementations but the performance penalties are an issue

**Comment by jbrockmendel** [comment]
> do you have a sense of how best to proceed here? My impression is that this will require two, if not three, steps:

These steps look about right.  This likely will take a while.

**Comment by znicholls** [comment]
Ok should I make a PR based on https://github.com/pandas-dev/pandas/compare/master...znicholls:handle-extension-arrays?

**Comment by jreback** [comment]
what is the status here? @znicholls

**Comment by znicholls** [comment]
> what is the status here?

Having rebased onto master, the issues raised here still exist (the new tests still fail).

Re-reading the issues and related PRs, I think https://github.com/pandas-dev/pandas/pull/35127#issuecomment-657391697 is the best summary i.e. the first step is identifying where we need the changes, the second step is implementing the changes (with my suggestion being less strict `is_list_like` and `is_scalar` to avoid performance hits but this might not be the best path forward once we've done step 1), the third step is finalisation (more or less).

I've had a go at doing step 1 (see https://github.com/pandas-dev/pandas/compare/master...znicholls:handle-extension-arrays). I'm waiting for advice on:

1. whether those identified spots to change make sense (alterations to `is_scalar` might also be needed but I'm struggling to work out why so have just left them out for now) 
1. whether going down the route of adding new implementations makes sense or whether there is a better route

**Comment by jreback** [comment]
@znicholls ok i see your comments above, not sure where this is going. I am not sure what your proposed implementation even looks like.

**Comment by znicholls** [comment]
The backstory is #35131. Hopefully that explains why we are looking to make such a change? (@jorisvandenbossche just double checking no progress has been made elsewhere that I've missed?)

A concern about performance impacts was raised in this comment https://github.com/pandas-dev/pandas/pull/35127#issuecomment-653954875. After some discussion, we arrived at https://github.com/pandas-dev/pandas/pull/35127#issuecomment-657391697. 

For completeness, we also did some checks about performance penalties, e.g. https://github.com/pandas-dev/pandas/pull/35127#issuecomment-654069720, https://github.com/pandas-dev/pandas/pull/35127#issuecomment-654530394 and https://github.com/pandas-dev/pandas/pull/35127#issuecomment-655178794. None of our solutions was perfect.

> I am not sure what your proposed implementation even looks like.

That makes sense, because I am also unsure. I've only got to step 1 of https://github.com/pandas-dev/pandas/pull/35127#issuecomment-657391697. Based on the 'hot spots' identified in https://github.com/pandas-dev/pandas/compare/master...znicholls:handle-extension-arrays, I am looking for advice about whether we could use less strict (potentially slower) implementations of `is_list_like` in these places.

If we can use less strict implementations, my only proposed implementation is the changes made in this PR i.e. https://github.com/pandas-dev/pandas/compare/master...znicholls:non-iterable-duck-arrays. In short, the less strict implementation would check for objects which are numpy-style zero-dimensional arrays with `not (hasattr(obj, "ndim") and obj.ndim == 0)` rather than `not (util.is_array(obj) and obj.ndim == 0)` because the latter only works for numpy arrays and doesn't work for **numpy-style** arrays (such as Pint).

**Comment by jreback** [comment]
@znicholls what is the status here

**Comment by znicholls** [comment]
> what is the status here

I would greatly appreciate your (or another pandas developer) on https://github.com/pandas-dev/pandas/pull/35127#issuecomment-757555672. I'm not sure I can sum it up any more succinctly than what is in that comment.

**Comment by jreback** [comment]
> If we can use less strict implementations, my only proposed implementation is the changes made in this PR i.e. master...znicholls:non-iterable-duck-arrays. In short, the less strict implementation would check for objects which are numpy-style zero-dimensional arrays with not (hasattr(obj, "ndim") and obj.ndim == 0) rather than not (util.is_array(obj) and obj.ndim == 0) because the latter only works for numpy arrays and doesn't work for numpy-style arrays (such as Pint).

i think this is reasonable. of course asv's will tell us whether its an acceptable perf hit. trying to balance additional complexity (seems minor in this case) with perf vs enhancements.

**Comment by znicholls** [comment]
> i think this is reasonable. of course asv's will tell us whether its an acceptable perf hit. trying to balance additional complexity (seems minor in this case) with perf vs enhancements.

Ok great thank you, I will start on that path then.

**Comment by znicholls** [comment]
@jreback I have tried to start small with #39790, if you could take a look that would be great

**Comment by mroeschke** [comment]
Thanks for the PR draft, but it appears this draft has stalled in development. Probably best if this change starts in a smaller PR with continuation of https://github.com/pandas-dev/pandas/pull/39790 (let us know if you'd like to continue in that PR. Closing.

**Comment by burnpanck** [comment]
What exactly is it that is missing here? There seems to be quite a number of issues in many downstream projects, all stalling on #35131. I myself would love to see pint-pandas move forward and would be willing to picking up this PR.

If I understand correctly, the main concern with the proposed solution here is the slight performance regression. To me, that then asks for performance re-optimisation rather than introducing two different versions of `is_list_like` that do slightly different things - the latter is calling for all kinds of subtle bugs. The size of the diff in #39830 (2500 changed lines) with the solution presented here (150 changed lines) also speaks for this approach here.

As for the performance re-optimisation: It seems there is no regression on actual numpy arrays, only on real lists. Under the assumption that real lists are the only relevant non-array types, we might be able to get by with a simple short-cut test for actual lists. Would an iteration on this PR along these lines be appreciated?

**Comment by jreback** [comment]
@burnpanck like anything in pandas a proper well formed and tested PR would be considered

**Comment by znicholls** [comment]
> The size of the diff in #39830 (2500 changed lines) with the solution presented here (150 changed lines) also speaks for this approach here.

I just rebased #39830 so you can see the real diff. It's only 59 lines and only 1 in the actual codebase (the rest are all tests). https://github.com/pandas-dev/pandas/compare/master...znicholls:update-is-list-like?expand=1

Thanks for picking up @burnpanck, as you can tell I had a few goes but always hit a wall at some point

**Pull request closed without merging**